src/feruchemy/java/leaf/cosmere/feruchemy/client/utils/FeruchemyChargeThread.java

```diff
@@ -0,0 +1,153 @@
+package leaf.cosmere.feruchemy.client.utils;
+
+import leaf.cosmere.api.CosmereAPI;
+import leaf.cosmere.api.Metals;
+import leaf.cosmere.common.items.ChargeableMetalCurioItem;
+import leaf.cosmere.feruchemy.common.itemgroups.FeruchemyItemGroups;
+import net.minecraft.client.Minecraft;
+import net.minecraft.world.item.ItemStack;
+import top.theillusivec4.curios.api.CuriosApi;
+
+import java.util.*;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class FeruchemyChargeThread implements Runnable {
+    private static final Lock lock = new ReentrantLock();
+    private static FeruchemyChargeThread INSTANCE;
+    private static final HashMap<Metals.MetalType, Double> feruchemyChargeMap = new HashMap<>();
+    static Thread t;
+    static boolean isStopping = false;
+
+    public static FeruchemyChargeThread getInstance()
+    {
+        if (INSTANCE == null)
+        {
+            INSTANCE = new FeruchemyChargeThread();
+        }
+
+        return INSTANCE;
+    }
+
+    public HashMap<Metals.MetalType, Double> getCharges()
+    {
+        lock.lock();
+        try {
+            HashMap<Metals.MetalType, Double> retVal = new HashMap<>(feruchemyChargeMap);
+            lock.unlock();
+            return retVal;
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+            lock.unlock();
+        }
+        return new HashMap<>();
+    }
+
+    public void start()
+    {
+        if (t == null || isStopping)
+        {
+            CosmereAPI.logger.info("Feruchemy thread started");
+            t = new Thread(this, "feruchemy_enumerator_thread");
+            isStopping = false;
+            t.start();
+        }
+    }
+
+    public void stop()
+    {
+        if (t != null && !isStopping)
+        {
+            isStopping = true;
+        }
+    }
+
+    @Override
+    public void run() {
+        Minecraft mc = Minecraft.getInstance();
+
+        while (!isStopping)
+        {
+            // no serverside action, unloaded levels, or non-existent players allowed >:(
+            if (mc.level == null || mc.player == null || !mc.level.isClientSide)
+            {
+                stop();
+                break;
+            }
+
+            if (mc.player.tickCount % 2 != 0)
+            {
+                try {
+                    Thread.sleep(50);       // 20 ticks per 1000ms, 1000/20 = 50, rest for 1 tick
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                continue;
+            }
+
+            // hashmap to keep track of each metal's f-charge in the inventory
+            final HashMap<Metals.MetalType, Double> metalmindCharges = new HashMap<>();
```

Comment:
We should probably save some garbage collection by reusing memory where possible, especially when we know we'll be needing it again every other tick

Reply:
Moved to the start of the thread instead, as we only need it while the thread is running